pyfar/dsp/dsp.py

```diff
@@ -227,6 +227,202 @@ def spectrogram(signal, dB=True, log_prefix=20, log_reference=1,
+            "The parameter shape has to be 'symmetric', 'left' or 'right'.")
+    if not isinstance(truncate, bool):
+        raise TypeError("The parameter truncate has to be of type: bool.")
+    if not isinstance(length, (list, type(None))):
+        raise TypeError(
+            "The parameter length has to be of type list or None.")
+
+    # Check length
+    if length is not None:
+        if length != sorted(length):
+            raise ValueError("Values in length need to be in ascending order.")
+        # Convert to samples
+        if unit == 's':
+            length = [round(li*signal.sampling_rate) for li in length]
+        elif unit == 'ms':
+            length = [round(li*signal.sampling_rate/1e3) for li in length]
+        elif unit == 'samples':
+            length = [int(li) for li in length]
+        else:
+            raise ValueError(f"unit is {unit} but has to be"
+                             f" 'samples', 's' or 'ms'.")
+        # Check window size
+        if length[-1] > signal.n_samples:
+            raise ValueError(
+                "Values in length require window to be longer than signal.")
+
+    # Create window
+    # win_start and win_stop define the first and last sample of the window
+    if length is None:
+        win_samples = signal.n_samples
+        win = sgn.windows.get_window(window, win_samples, fftbins=False)
+        win_start = 0
+        win_stop = signal.n_samples-1
+    elif len(length) == 2:
+        if shape == 'symmetric':
+            win_samples = length[1]-length[0]+1
+            win = sgn.windows.get_window(window, win_samples, fftbins=False)
+            win_start = length[0]
+            win_stop = length[1]
+        else:
+            fade_samples = int(2*(length[1]-length[0]))
+            fade = sgn.windows.get_window(window, fade_samples, fftbins=False)
+            if shape == 'left':
+                win = np.ones(signal.n_samples-length[0])
+                win[0:length[1]-length[0]] = fade[:int(fade_samples/2)]
+                win_start = length[0]
+                win_stop = signal.n_samples-1
+            if shape == 'right':
+                win = np.ones(length[1]+1)
+                win[length[0]+1:] = fade[int(fade_samples/2):]
+                win_start = 0
+                win_stop = length[1]
+    elif len(length) == 4:
+        fade_in_samples = int(2*(length[1]-length[0]))
+        fade_in = sgn.windows.get_window(
+            window, fade_in_samples, fftbins=False)
+        fade_in = fade_in[:int(fade_in_samples/2)]
+        fade_out_samples = int(2*(length[3]-length[2]))
+        fade_out = sgn.windows.get_window(
+            window, fade_out_samples, fftbins=False)
+        fade_out = fade_out[int(fade_out_samples/2):]
+        win = np.ones(length[-1]-length[0]+1)
+        win[0:length[1]-length[0]] = fade_in
+        win[length[2]-length[0]+1:length[3]-length[0]+1] = fade_out
+        win_start = length[0]
+        win_stop = length[3]
```

Comment:
This is bit of redundant code here: Couldn't this be implemented as a recursive function using the first two cases, simply multiplying left and right sided windows in the end for the last case?

Reply:
The problem is that the right slope depends on length[0] and length[1] for shape='right', but on length[2] and length[3] for the symmetric case. The only alternative I see would be to create a right slope with length[-2] and length[-1] in all cases, but in the case without fade out this would cause obsolete additional computational effort.

Reply:
I think computational effort is not really something we need to worry about in this function. I would go for the most readable solution. I was thinking about something like this (my own very simple window function I've been using for my personal projects, just as an example):

``` python
def _half_sided_window(interval, n_samples):
    window_length = np.int(np.diff(interval))
    window = sps.hann(2*np.abs(window_length))
    if np.sign(window_length) == -1:
        window_left = window[:np.abs(window_length)]
        window_full = np.concatenate((
            np.zeros(interval[1]),
            window_left,
            np.ones(n_samples-interval[0])))

    else:
        window_right = window[np.int(np.abs(window_length))::]
        window_full = np.concatenate((
            np.ones(interval[0]),
            window_right,
            np.zeros(n_samples-interval[1])))

    return window_full


def time_window(data, intervals, window='hanning', samples=True):
    n_samples = data.shape[-1]

    intervals = np.asarray(intervals)

    if np.size(intervals) == 2:
        window_full = _half_sided_window(intervals, n_samples)
    elif np.size(intervals) == 4:
        window_left = _half_sided_window(intervals[:2], n_samples)
        window_right = _half_sided_window(intervals[2:], n_samples)
        window_full = window_left * window_right

    data = data.copy() * np.broadcast_to(window_full, data.shape)

    return data
```

Reply:
Would be fine for me, but I don't have a strong preference for either.

Reply:
Wrote private functions to increase readability